### DISCUSSION TITLE
Has memory management changed in X11 recently?

### DISCUSSION THREAD

**Discussion by tsujan** [lxqt/lxqt#2321] [Q&A] [answered]
I don't know if this is about KWin, X11 or something else, but my RAM usage has been increased recently (I don't know after which upgrade in Manjaro/Arch), while I run no new app. KWin and Xorg memory usages seem normal, and I can't find anything unusual in other apps either — LXQt apps are definitely OK.

The suspicious RAM usage may start in the 2nd or 3rd day of working. If I log out and log into Wayfire, the RAM usage will remain normal after 3 days. So, I think this may be related to X11.

Has anyone noticed a similar behavior?

**Comment by stefonarch** [comment]
I've had a similar issue but with wayfire -  ran out of memory after some day and had a full swap, but didn't find the culprit. Tested X11 too but not over 3 days. Suspected quiterss and firefox.

**Reply by tsujan** [reply]
> I've had a similar issue but with wayfire - ran out of memory after some day and had a full swap

Oh, that's not good. I haven't found a memory leak in Wayfire itself, but third-party (or extra) plugins may have leaks.

It doesn't seem so bad here, under an ordinary X11 LXQt session, but I think something has changed after an upgrade. On the one hand, the memory usages of Qt6 apps have decreased (I didn't check Qt5 apps); on the other hand, the total memory usage is greater than before, after 2-3 days of working under a session!

**Reply by stefonarch** [reply]
I've no idea, I hadn't updated for weeks (except firefox which updated itself if I remember well). 
Now it seems gone, but the desktop blocked twice when it happened.

**Comment by darkshram** [comment]
I will be a good idea to post exactly which versions of X11 components are used in system with memory issues. LibX11 1.8.1 had several issues with [Xfce](https://gitlab.xfce.org/xfce/xfce4-settings/-/issues/369) and [fvwm3](https://bugs.gentoo.org/862115). This release of libx11 included fixes that also [made evident some bugs](https://duckduckgo.com/?q=libx11+1.8.1+bug) in some applications (like the one in xfce4-settings), so I would not discard something related or similar with other X11 related applications.

For my Linux distribution (ALDOS), there are no issues (**yet**) with LXQt. A freshly booted system still uses ~200 MB RAM after upgrading to LXQt 1.2.0.

  - libX11 1.8.1 (this version [requires](https://bugs.freedesktop.org/show_bug.cgi?id=3104) `/var/cache/libx11/compose` or `~/.compose-cache` to exist for compose caching)
  - libXext 1.3.5
  - xorg-x11-server-Xorg 1.20.14 (we chose to not upgrade yet to 2022.x because the legacy Nvidia drivers support).
  - glib 2.68
  - Qt 5.15.4
  - As default, we use openbox 3.6.1 as wm for LXQt.
  - Lightdm+lightm-gtk-greeter as default display manager
  - Xscreensaver 6.04

**Reply by tsujan** [reply]
> LibX11 1.8.1 had several issues with Xfce and fvwm3.

@darkshram, thanks! That's a useful info.

I have `libx11` 1.8.1 too (`libx11` 1.8.1-3 from Arch), with `libxext` 1.3.5 and `xorg-server` 21.1.4.

> there are no issues (yet) with LXQt.

I don't think there can be any issue with LXQt itself because LXQt doesn't have a WM or compositor. I don't see any problem with KWin either (under LXQt).

It's just that the total memory usage seems higher than before, after 5 days, while the memory usages of individual apps seem like before. This may be good for some but isn't normal for me:
```
$ free
               total        used        free      shared  buff/cache   available
Mem:        16207176     1597804     3377560     2033188    11231812    12234544
Swap:       16383996          20    16383976
```

**Comment by tsujan** [comment]
IMO, there should be a problem (or regression) in this version of X11. The memory usage grows gradually but indefinitely, day after day. It doesn't happen under Wayfire, where I use LXQt+Qt apps.

I'll wait for the next version of X11.

**Reply by darkshram** [reply]
Seems like a specific bug for Manjaro/Arch. Obviously there is a memory leak somewhere. I have not been able to replicate it in my Linux distribution (uptime: ` 18:52:14 up 15 days,  7:09,  2 users,  load average: 0.46, 0.35, 0.28
`), neither Fedora with Xfce, LXQt, MATE, Cinnamon ort KDE Plasma. I am intrigued because as main packager for ALDOS (basically, Fedora without SystemD), and find this issue relevant. 

Are you using only the FLOSS Xorg drivers or any of the proprietary Nvidia/AMD ones?

Could you please post the output of `ps aux` while system is using too much memory?

[I have not found](https://gitlab.freedesktop.org/xorg/lib/libx11/-/issues) any current memory leak related issues for libX11. So my best guess with the information to this point would be Xorg server, display manager or any other background process. I would not discard a SystemD-related issue as the culprit.

In case you find it useful, I built my RPM packages with the patch attached to this message for libX11 1.8.1.
[dont-forward-keycode-0.zip](https://github.com/lxqt/lxqt/files/9966096/dont-forward-keycode-0.zip). This patch was taken from [Fedora](https://src.fedoraproject.org/rpms/libX11) and [is not included in arch](https://github.com/archlinux/svntogit-packages/blob/packages/libx11/trunk/PKGBUILD) (probably is a Fedora/RHEL specific patch, but please look at the code).

Aside for this patch, everything else is the standard `./configure; make; make install` (as macros):

```
%prep
%autosetup -p1 -n %{tarball}-%{version}

%build
autoreconf -v --install --force
%configure --disable-silent-rules --disable-static
%make_build

%install
%make_install

# create/own compose cache dir
mkdir -p %{buildroot}/var/cache/libX11/compose

# We intentionally don't ship *.la files
find %{buildroot} -type f -name '*.la' -delete
```

My suggestions:

a) downgrade to 1.8.0 to test if it is related to 1.8.1. As I said before, fixes en 1.8.1 made bugs in some applications and other X11 components to be evident. So, is a little complicated to classify any related bug to libX11 1.8.1 as a true _regression_.

b) Keep libX11 1.8.1 but downgrade Xorg-server to 1.20.14. This would mean to also rebuild the installed xf86-drv-* packages (to avoid API breakage).

**Reply by tsujan** [reply]
> Are you using only the FLOSS Xorg drivers or any of the proprietary Nvidia/AMD ones?

I use Intel + modesettings + latest LTS kernel.

> my best guess with the information to this point would be Xorg, display manager or any other background process.

I first suspected KWin, but its memory usage seems normal. Xorg seems OK too. Not sure of SDDM though.

Needing a stable and up-to-date system for coding, I can't experiment by downgrading fundamental packages; hence, waiting for upgrades. The problem isn't disastrous and, for now, I've avoided it completely by working under Wayland.

Thanks for the detailed info and suggestions!

**Comment by tsujan** [comment]
OK, I think I found the cause: it had something to do with KWin's compositing. Once I disabled and enabled it again (by chance), the memory usage became normal, even after a logout and login.

It can be related to a backward-incompatible setting of KWin ≥ 5.25 and may not happen with newly installed systems. But there should be a bug somewhere in the new code. KDE devs fixed some serious issues in the point releases of 5.25, but 5.26 still has bugs that didn't exist before.

**Comment by tsujan** [comment, marked as answer]
Although the problem became milder after restarting Kwin's compositing, it didn't disappear completely.

The memory leak of KWin 5.26 (and, maybe, older versions) is here: KDE System Settings → Worksace behavior → Screen Edges → Maximize, Tile. Those check-boxes correspond to `ElectricBorderMaximize` and `ElectricBorderTiling` in `~/.config/kwinrc`.

When these settings are enabled, each window tiling *with mouse* will reduce the available memory by a few megabytes. Perhaps an allocated memory isn't released after DND.

Window tiling with shortcut isn't affected.

EDIT: KWin's code may have other holes too. I'm just sure of `ElectricBorderMaximize` and `ElectricBorderTiling`.

**Comment by stefonarch** [comment]
I just happened to fill my ram too but on wayfire. My suspect is still firefox, which left over some webcontent  processes  also after closing, but killing those didn't free anything, had to restart session.

**Reply by tsujan** [reply]
@stefonarch 
I think we've encountered totally different problems. What I see is related to X11 + KWin; what you see is related to Wayland + Wayfire.

Interestingly enough, before I found out how to avoid KWin's memory leak under X11 (by disabling "electric borders"), I used Wayfire + LXQt to have a stable memory usage, while the same Wayfire caused a memory leak in your case!

**Reply by stefonarch** [reply]
As I said I suspect the browser here, and saw similar issue on X11, but as I used it less I can't really tell. No electric borders though.

**Reply by tsujan** [reply]
> As I said I suspect the browser here

Are you sure thar it isn't related to `xdg-desktop-portal`? I don't have it here.

**Reply by tsujan** [reply]
Let me be more specific about my experience:

I obsessively monitored memory usages of xorg, kwin, thunderbird, lxqt-panel and pcmanfm-qt, because they should grow a little with time. I also checked Firefox, Chromium, and a few other heavy apps: they released their memory after being closed.

Before disabling KWin's "electric borders", the growth in the RAM usage was 50-100 MiB *more than* the total growth in the memory usages of those programs, every few hours. In other words, I couldn't account for 50-100 MiB in 2-3 hours. After disabling "electric borders", everything looked normal — or, at least, much better.

The reason was that I have a habit of tiling windows by dragging them to screen edges/corners with mouse. That was exactly the weak point of KWin: it wasn't reflected by KWin's memory usage but decreased the available memory. From where? I don't know; it wasn't from Xorg.

**Reply by stefonarch** [reply]
> > As I said I suspect the browser here
> 
> Are you sure thar it isn't related to `xdg-desktop-portal`? I don't have it here.

As I managed to break it's function it's not installed atm.

**Reply by stefonarch** [reply]
Looks I found the culprit in wayland, it's cinny, a matrix client. Element (my preferred one) has a bug when switching between X11 and wayland, not connecting anymore).

**Reply by tsujan** [reply]
> Looks I found the culprit in wayland

Wayland's peculiarities aside, as far as I've seen, the serious problems apps may have with Wayland are usually caused by bugs in those apps.
